drivers/ethernet/eth_numaker.c

```diff
@@ -0,0 +1,807 @@
+}
+
+static int mdio_read(synopGMACdevice *gmacdev, uint32_t addr, uint32_t reg)
+{
+	uint16_t data;
+
+	synopGMAC_read_phy_reg((u32 *)gmacdev->MacBase, addr, reg, &data);
+	return data;
+}
+
+static int numaker_eth_link_ok(synopGMACdevice *gmacdev)
+{
+	/* first, a dummy read to latch */
+	mdio_read(gmacdev, eth_phy_addr, MII_BMSR);
+	if (mdio_read(gmacdev, eth_phy_addr, MII_BMSR) & BMSR_LSTATUS) {
+		return 1;
+	}
+	return 0;
+}
+
+static int reset_phy(synopGMACdevice *gmacdev)
+{
+	uint16_t reg;
+	uint32_t delay_cnt;
+	int status = 0;
+	bool ret;
+
+	mdio_write(gmacdev, eth_phy_addr, MII_BMCR, BMCR_RESET);
+
+	delay_cnt = 200000U;
+	ret = WAIT_FOR(!(mdio_read(gmacdev, eth_phy_addr, MII_BMCR) & BMCR_RESET),
+					delay_cnt, k_msleep(1));
+
+	if (ret == false) {
+		LOG_DBG("Reset phy failed");
+		return -ENODATA;
+	}
+
+	delay_cnt = 3000000U;
+	ret = WAIT_FOR(numaker_eth_link_ok(gmacdev), delay_cnt, k_msleep(1));
+
+	if (ret) {
+		gmacdev->LinkState = LINKUP;
+		LOG_DBG("Link Up");
+	}
+
+	if (ret == false) {
+		gmacdev->LinkState = LINKDOWN;
+		LOG_DBG("Link Down");
+		return -ENODATA;
+	}
+
+	mdio_write(gmacdev, eth_phy_addr, MII_ADVERTISE,
+		   ADVERTISE_CSMA | ADVERTISE_10HALF | ADVERTISE_10FULL | ADVERTISE_100HALF |
+			   ADVERTISE_100FULL);
+
+	reg = mdio_read(gmacdev, eth_phy_addr, MII_BMCR);
+	mdio_write(gmacdev, eth_phy_addr, MII_BMCR, reg | BMCR_ANRESTART);
+
+	delay_cnt = 3000000U;
+	ret = WAIT_FOR((mdio_read(gmacdev, eth_phy_addr, MII_BMSR) &
+					(BMSR_ANEGCOMPLETE | BMSR_LSTATUS)) ==
+					(BMSR_ANEGCOMPLETE | BMSR_LSTATUS),
+					delay_cnt, k_msleep(1));
+
+	if (ret == false) {
+		LOG_DBG("AN failed. Set to 100 FULL");
+		synopGMAC_set_full_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 1); /* Set mode 1: 100Mbps; 2: 10Mbps */
+		return -ENODATA;
+	}
+	reg = mdio_read(gmacdev, eth_phy_addr, MII_LPA);
+	if (reg & ADVERTISE_100FULL) {
+		LOG_DBG("100 full");
+		gmacdev->DuplexMode = FULLDUPLEX;
+		gmacdev->Speed = SPEED100;
+		synopGMAC_set_full_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 1); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	} else if (reg & ADVERTISE_100HALF) {
+		LOG_DBG("100 half");
+		gmacdev->DuplexMode = HALFDUPLEX;
+		gmacdev->Speed = SPEED100;
+		synopGMAC_set_half_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 1); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	} else if (reg & ADVERTISE_10FULL) {
+		LOG_DBG("10 full");
+		gmacdev->DuplexMode = FULLDUPLEX;
+		gmacdev->Speed = SPEED10;
+		synopGMAC_set_full_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 2); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	} else {
+		LOG_DBG("10 half");
+		gmacdev->DuplexMode = HALFDUPLEX;
+		gmacdev->Speed = SPEED10;
+		synopGMAC_set_half_duplex(gmacdev);
+		synopGMAC_set_mode(NUMAKER_GMAC_INTF, 2); /* Set mode 1: 100Mbps; 2: 10Mbps */
+	}
+
+	LOG_INF("PHY ID 1:0x%x", mdio_read(gmacdev, eth_phy_addr, MII_PHYSID1));
+	LOG_INF("PHY ID 2:0x%x", mdio_read(gmacdev, eth_phy_addr, MII_PHYSID2));
```

Comment:
would it make more sense to move these up? print the id before any write, so that if any other write fail you would at least have an ID that can tell you if something is off